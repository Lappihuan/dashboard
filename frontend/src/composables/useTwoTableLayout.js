//
// SPDX-FileCopyrightText: 2025 SAP SE or an SAP affiliate company and Gardener contributors
//
// SPDX-License-Identifier: Apache-2.0
//

import {
  computed,
  useTemplateRef,
} from 'vue'
import { useElementSize } from '@vueuse/core'

export function useTwoTableLayout ({
  containerRefName = 'container', // Element containing two vuetify data tables
  itemHeight, // height of table items
  tableHeaderHeight = 40, // height of table headers
  tableFooterHeight = 37, // height of table footers
  yMargins = 90, // combined vertical margins
  noDataHeight = 120, // reserved space in case of no data
  firstTableItemCount,
  secondTableItemCount,
}) {
  const containerRef = useTemplateRef(containerRefName)
  const { height: containerHeight } = useElementSize(containerRef)

  const desiredFirstTableHeight = computed(() => {
    const contentHeight = firstTableItemCount.value * (itemHeight + 1)
    if (contentHeight === 0) {
      return noDataHeight
    }
    return contentHeight + tableHeaderHeight + tableFooterHeight
  })

  const desiredSecondTableHeight = computed(() => {
    const contentHeight = secondTableItemCount.value * (itemHeight + 1)
    if (contentHeight === 0) {
      return noDataHeight
    }
    return contentHeight + tableHeaderHeight + tableFooterHeight
  })

  const halfAvailableHeight = computed(() => {
    return containerHeight.value / 2 - yMargins
  })

  const firstTableStyles = computed(() => {
    const additionalHeight = halfAvailableHeight.value - desiredSecondTableHeight.value
    let availableHeight = halfAvailableHeight.value
    if (additionalHeight > 0) {
      availableHeight = availableHeight + additionalHeight
    }
    const height = Math.min(desiredFirstTableHeight.value, availableHeight)

    return {
      height: `${height}px`,
    }
  })

  const secondTableStyles = computed(() => {
    const additionalHeight = halfAvailableHeight.value - desiredFirstTableHeight.value
    let availableHeight = halfAvailableHeight.value
    if (additionalHeight > 0) {
      availableHeight = availableHeight + additionalHeight
    }
    const height = Math.min(desiredSecondTableHeight.value, availableHeight)
    return {
      height: `${height}px`,
    }
  })
  return {
    firstTableStyles,
    secondTableStyles,
  }
}
