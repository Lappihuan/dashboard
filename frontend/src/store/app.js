//
// SPDX-FileCopyrightText: 2023 SAP SE or an SAP affiliate company and Gardener contributors
//
// SPDX-License-Identifier: Apache-2.0
//

import {
  defineStore,
  acceptHMRUpdate,
} from 'pinia'
import {
  ref,
  watch,
  toRef,
  computed,
} from 'vue'
import LuigiClient from '@luigi-project/client'

import { useLogger } from '@/composables/useLogger'

import moment from '@/utils/moment'
import { errorDetailsFromError } from '@/utils/error'

export const useAppStore = defineStore('app', () => {
  const logger = useLogger()

  const ready = ref(false)
  const sidebar = ref(true)
  const redirectPath = ref(null)
  const loading = ref(false)
  const alert = ref(null)
  const location = ref(moment.tz.guess())
  const timezone = ref(moment().format('Z'))
  const focusedElementId = ref(null)
  const splitpaneResize = ref(0)
  const fromRoute = ref(null)
  const routerError = ref(null)
  const luigiContext = ref(null)

  const isInIframe = computed(() => window.self !== window.top)

  if (isInIframe.value) {
    logger.debug('Registering listener for Luigi context initialization and context updates')
    LuigiClient.addInitListener(context => setLuigiContext(context))
    LuigiClient.addContextUpdateListener(context => setLuigiContext(context))
  }

  function setLuigiContext (value) {
    luigiContext.value = value
  }

  const accountId = computed(() => luigiContext.value?.accountId)

  function getLuigiContext () {
    if (!isInIframe.value) {
      return Promise.resolve(null)
    }
    if (luigiContext.value !== null) {
      return Promise.resolve(luigiContext.value)
    }
    return new Promise(resolve => {
      const timeout = 3000
      const timeoutId = setTimeout(() => {
        unwatch()
        logger.error('The initialization of the Luigi Client has timed out after %d milliseconds', timeout)
        resolve(null)
      }, timeout)
      const unwatch = watch(luigiContext, context => {
        if (context !== null) {
          clearTimeout(timeoutId)
          unwatch()
          resolve(context)
        }
      }, {
        immediate: true,
      })
    })
  }

  function setRoute (route) {
    if (isInIframe.value) {
      const pathname = toRef(route, 'path')
      watch(pathname, value => {
        if (value) {
          LuigiClient.linkManager().fromVirtualTreeRoot().withoutSync().navigate(value)
        }
      }, {
        immediate: true,
      })
    }
  }

  function updateSplitpaneResize () {
    splitpaneResize.value = Date.now()
  }

  function setAlert (value) {
    alert.value = value
  }

  function setAlertWithType (type, value) {
    const alert = {
      type,
    }
    if (typeof value === 'string') {
      alert.message = value
    } else if (value) {
      const { message = '', title } = value
      alert.message = value.response
        ? errorDetailsFromError(value).detailedMessage
        : message
      if (title) {
        alert.title = title
      }
    } else {
      alert.message = ''
    }
    setAlert(alert)
  }

  function setError (value) {
    setAlertWithType('error', value)
  }

  function setSuccess (value) {
    setAlertWithType('success', value)
  }

  function setRouterError (value) {
    routerError.value = value
  }

  return {
    ready,
    sidebar,
    redirectPath,
    loading,
    alert,
    location,
    timezone,
    focusedElementId,
    splitpaneResize,
    fromRoute,
    isInIframe,
    setRoute,
    routerError,
    accountId,
    luigiContext,
    getLuigiContext,
    updateSplitpaneResize,
    setAlert,
    setError,
    setSuccess,
    setRouterError,
  }
})

if (import.meta.hot) {
  import.meta.hot.accept(acceptHMRUpdate(useAppStore, import.meta.hot))
}
