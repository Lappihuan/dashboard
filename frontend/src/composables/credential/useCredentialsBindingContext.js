//
// SPDX-FileCopyrightText: 2024 SAP SE or an SAP affiliate company and Gardener contributors
//
// SPDX-License-Identifier: Apache-2.0
//

import {
  ref,
  shallowRef,
  computed,
} from 'vue'

import { useAuthzStore } from '@/store/authz'

import { cleanup } from '@/composables/helper'
import { useObjectMetadata } from '@/composables/useObjectMetadata'

import cloneDeep from 'lodash/cloneDeep'
import get from 'lodash/get'
import isEqual from 'lodash/isEqual'
import set from 'lodash/set'

export function useCredentialsBindingContext (options = {}) {
  const {
    authzStore = useAuthzStore(),
  } = options

  const initialCredentialsBindingManifest = shallowRef(null)

  const normalizedInitialCredentialsBindingManifest = computed(() => {
    const object = cloneDeep(initialCredentialsBindingManifest.value)
    return normalizeCredentialsBindingManifest(object)
  })

  const credentialsBindingManifest = ref({})

  function normalizeCredentialsBindingManifest (value) {
    const object = Object.assign({
      apiVersion: 'security.gardener.cloud/v1alpha1',
      kind: 'CredentialsBinding',
      credentialsRef: {
        apiVersion: 'v1',
        kind: 'Secret',
      },
    }, value)
    return cleanup(object)
  }

  const normalizedCredentialsBindingManifest = computed(() => {
    const object = cloneDeep(credentialsBindingManifest.value)
    return normalizeCredentialsBindingManifest(object)
  })

  function setCredentialsBindingManifest (value) {
    initialCredentialsBindingManifest.value = value
    credentialsBindingManifest.value = cloneDeep(initialCredentialsBindingManifest.value)
  }

  function createCredentialsBindingManifest () {
    const namespace = get(options, ['namespace'], authzStore.namespace)
    credentialsBindingManifest.value = {
      metadata: {
        name: '',
        namespace,
      },
      provider: {
        type: '',
      },
      credentialsRef: {
        name: '',
        namespace,
      },
    }
    initialCredentialsBindingManifest.value = cloneDeep(credentialsBindingManifest.value)
  }

  const isCredentialsBindingDirty = computed(() => {
    return !isEqual(
      normalizedCredentialsBindingManifest.value,
      normalizedInitialCredentialsBindingManifest.value,
    )
  })

  const {
    name: credentialsBindingName,
    namespace: credentialsBindingNamespace,
  } = useObjectMetadata(credentialsBindingManifest)

  const credentialsBindingProviderType = computed({
    get () {
      return get(credentialsBindingManifest.value, ['provider', 'type'])
    },
    set (value) {
      set(credentialsBindingManifest.value, ['provider', 'type'], value)
    },
  })

  const credentialsBindingCredentialsRef = computed({
    get () {
      return get(credentialsBindingManifest.value, ['credentialsRef'])
    },
    set (value) {
      set(credentialsBindingManifest.value, ['credentialsRef'], value)
    },
  })

  return {
    bindingManifest: normalizedCredentialsBindingManifest,
    setBindingManifest: setCredentialsBindingManifest,
    createBindingManifest: createCredentialsBindingManifest,
    isBindingDirty: isCredentialsBindingDirty,
    bindingName: credentialsBindingName,
    bindingNamespace: credentialsBindingNamespace,
    bindingProviderType: credentialsBindingProviderType,
    bindingRef: credentialsBindingCredentialsRef,
  }
}
