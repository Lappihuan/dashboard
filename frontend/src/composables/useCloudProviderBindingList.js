//
// SPDX-FileCopyrightText: 2024 SAP SE or an SAP affiliate company and Gardener contributors
//
// SPDX-License-Identifier: Apache-2.0
//

import { computed } from 'vue'

import { useCredentialStore } from '@/store/credential'
import { useGardenerExtensionStore } from '@/store/gardenerExtension'

import { hasOwnCredential } from '@/utils'

import filter from 'lodash/filter'

export const useCloudProviderBindingList = (providerType, options = {}) => {
  const {
    credentialStore = useCredentialStore(),
    gardenerExtensionStore = useGardenerExtensionStore(),
  } = options

  return computed(() => {
    const isDnsSecret = gardenerExtensionStore.dnsProviderTypes.includes(providerType.value)
    return filter(credentialStore.cloudProviderBindingList, secretBinding => {
      if (secretBinding.provider?.type !== providerType.value) {
        return false
      }

      return isDnsSecret
        ? hasOwnCredential(secretBinding) // shared dns secret not supported by dns extension
        : true
    })
  })
}
