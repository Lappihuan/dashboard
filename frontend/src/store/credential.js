//
// SPDX-FileCopyrightText: 2023 SAP SE or an SAP affiliate company and Gardener contributors
//
// SPDX-License-Identifier: Apache-2.0
//

import {
  defineStore,
  acceptHMRUpdate,
} from 'pinia'
import {
  computed,
  reactive,
} from 'vue'

import { useApi } from '@/composables/useApi'

import { useAuthzStore } from './authz'
import { useAppStore } from './app'
import { useGardenerExtensionStore } from './gardenerExtension'
import { useCloudProfileStore } from './cloudProfile'

import filter from 'lodash/filter'
import set from 'lodash/set'
import get from 'lodash/get'

function namespaceNameKey ({ namespace, name }) {
  return `${namespace}/${name}`
}

export const useCredentialStore = defineStore('credential', () => {
  const api = useApi()
  const appStore = useAppStore()
  const authzStore = useAuthzStore()
  const gardenerExtensionStore = useGardenerExtensionStore()
  const cloudProfileStore = useCloudProfileStore()

  const state = reactive({
    secretBindings: {},
    secrets: {},
    credentialsBindings: {},
    workloadIdentities: {},
    quotas: {},
  })

  function $reset () {
    state.secretBindings = {}
    state.secrets = {}
    state.credentialsBindings = {}
    state.workloadIdentities = {}
    state.quotas = {}
  }

  async function fetchCredentials () {
    const namespace = authzStore.namespace
    try {
      const { data: { secretBindings, secrets, credentialsBindings, workloadIdentities, quotas } } = await api.getCloudProviderCredentials(namespace)
      _setCredentials({ secretBindings, secrets, credentialsBindings, workloadIdentities, quotas })
    } catch (err) {
      $reset()
      throw err
    }
  }

  function _setCredentials ({ secretBindings, secrets, credentialsBindings, workloadIdentities, quotas }) {
    $reset()

    secretBindings?.forEach(item => {
      const key = namespaceNameKey(item.metadata)
      set(state.secretBindings, [key], item)
    })

    secrets?.forEach(item => {
      const key = namespaceNameKey(item.metadata)
      set(state.secrets, [key], item)
    })

    credentialsBindings?.forEach(item => {
      const key = namespaceNameKey(item.metadata)
      set(state.credentialsBindings, [key], item)
    })

    workloadIdentities?.forEach(item => {
      const key = namespaceNameKey(item.metadata)
      set(state.workloadIdentities, [key], item)
    })

    quotas?.forEach(item => {
      const key = namespaceNameKey(item.metadata)
      set(state.quotas, [key], item)
    })
  }

  const cloudProviderBindingList = computed(() => {
    const decorateBinding = (binding, kind) => {
      const isSecretBinding = kind === 'SecretBinding'
      const isCredentialsBinding = kind === 'CredentialsBinding'

      if (isSecretBinding) {
        const secret = getSecret(binding.secretRef)
        Object.defineProperty(binding, '_secret', {
          value: secret,
          configurable: true,
        })
        Object.defineProperty(binding, '_secretName', {
          value: binding.secretRef.name,
          configurable: true,
        })
      }
      if (isCredentialsBinding) {
        if (binding.credentialsRef.kind === 'Secret') {
          const secret = getSecret(binding.credentialsRef)
          Object.defineProperty(binding, '_secret', {
            value: secret,
            configurable: true,
          })
          Object.defineProperty(binding, '_secretName', {
            value: binding.credentialsRef.name,
            configurable: true,
          })
        }
        if (binding.credentialsRef.kind === 'WorkloadIdentity') {
          const workloadIdentity = getWorkloadIdentity(binding.credentialsRef)
          Object.defineProperty(binding, '_workloadIdentity', {
            value: workloadIdentity,
            configurable: true,
          })
        }
      }

      Object.defineProperty(binding, '_isSecretBinding', {
        value: isSecretBinding,
        configurable: true,
      })

      Object.defineProperty(binding, '_isCredentialsBinding', {
        value: isCredentialsBinding,
        configurable: true,
      })

      const quotas = (binding.quotas || [])
        .map(quota => get(state.quotas, namespaceNameKey(quota)))
        .filter(Boolean)

      Object.defineProperty(binding, '_quotas', {
        value: quotas,
        configurable: true,
      })

      return binding
    }

    const secretBindings = Object.values(state.secretBindings).map(binding => decorateBinding(binding, 'SecretBinding'))
    const credentialsBindings = Object.values(state.credentialsBindings).map(binding => decorateBinding(binding, 'CredentialsBinding'))

    return [...secretBindings, ...credentialsBindings]
  })

  const quotaList = computed(() => {
    return Object.values(state.quotas)
  })

  // TODO support credentials bindings
  async function createCredential (params) {
    const { data: { secretBinding, secret } } = await api.createCloudProviderCredential({ secretBinding: params.secretBinding, secret: params.secret })
    _updateCloudProviderCredential({ secretBinding, secret })
    appStore.setSuccess(`Cloud Provider credential ${secretBinding.metadata.name} created`)
  }

  // TODO support credentials bindings
  async function updateCredential (params) {
    const { data: { secretBinding, secret } } = await api.updateCloudProviderCredential({ secretBinding: params.secretBinding, secret: params.secret })
    _updateCloudProviderCredential({ secretBinding, secret })
    appStore.setSuccess(`Cloud Provider credential ${secretBinding.metadata.name} updated`)
  }

  // TODO support credentials bindings
  async function deleteCredential (name) {
    const namespace = authzStore.namespace

    await api.deleteCloudProviderCredential({ namespace, name })
    await fetchCredentials()
    appStore.setSuccess(`Cloud Provider credential ${name} deleted`)
  }

  const infrastructureBindingList = computed(() => {
    return filter(cloudProviderBindingList.value, binding => {
      return cloudProfileStore.sortedProviderTypeList.includes(binding.provider?.type)
    })
  })

  const dnsBindingList = computed(() => {
    return filter(cloudProviderBindingList.value, binding => {
      return gardenerExtensionStore.dnsProviderTypes.includes(binding.provider?.type) &&
        !!binding._secret // dns extension currently supports secrets only (no bindings)
    })
  })

  function getSecret ({ namespace, name }) {
    return get(state.secrets, [namespaceNameKey({ namespace, name })])
  }

  function getWorkloadIdentity ({ namespace, name }) {
    return get(state.workloadIdentities, [namespaceNameKey({ namespace, name })])
  }

  function getSecretBinding ({ namespace, name }) {
    return get(state.secretBindings, [namespaceNameKey({ namespace, name })])
  }

  function getCredentialsBinding ({ namespace, name }) {
    return get(state.credentialsBindings, [namespaceNameKey({ namespace, name })])
  }

  // TODO support credentials bindings
  function _updateCloudProviderCredential ({ secretBinding, secret }) {
    const key = namespaceNameKey(secretBinding.metadata)
    set(state.secretBindings, [key], secretBinding)

    if (secret) {
      // technically speaking secret should always be there as we currently only support to create secret and secret binding together
      // however this might change in the future
      const key = namespaceNameKey(secret.metadata)
      set(state.secrets, [key], secret)
    }

    // no update logic for quotas as they currently cannot be updated using the dashboard
  }

  return {
    cloudProviderBindingList,
    quotaList,
    fetchCredentials,
    _setCredentials,
    updateCredential,
    createCredential,
    deleteCredential,
    infrastructureBindingList,
    dnsBindingList,
    getSecret,
    getSecretBinding,
    getCredentialsBinding,
    getWorkloadIdentity,
    $reset,
  }
})

if (import.meta.hot) {
  import.meta.hot.accept(acceptHMRUpdate(useCredentialStore, import.meta.hot))
}
