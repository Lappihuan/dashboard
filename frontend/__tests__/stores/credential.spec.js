//
// SPDX-FileCopyrightText: 2024 SAP SE or an SAP affiliate company and Gardener contributors
//
// SPDX-License-Identifier: Apache-2.0
//

import {
  setActivePinia,
  createPinia,
} from 'pinia'

import { useAppStore } from '@/store/app'
import { useAuthzStore } from '@/store/authz'
import { useCloudProfileStore } from '@/store/cloudProfile'
import { useCredentialStore } from '@/store/credential'
import { useGardenerExtensionStore } from '@/store/gardenerExtension'

import { useApi } from '@/composables/useApi'

import find from 'lodash/find'

const testNamespace = 'garden-test'
const awsSecretBindingName = 'aws-secretbinding'
const awsCredentialsBindingName = 'aws-credentialsbinding'
const awsSecretName = 'aws-secret'
const awsWorkloadIndentityName = 'aws-wlid'
const awsTrialSecretBindingName = 'aws-trial-secretbinding'
const awsTrialCredentialsBindingName = 'aws-trial-credentialsbinding'
const azureSecretBindingName = 'azure-secretbinding'
const azureCredentialsBindingName = 'azure-credentialsbinding'
const awsWorkloadIdentitiyCredentialsBindingName = 'aws-wlid-credentialsbinding'

describe('stores', () => {
  describe('credential', () => {
    let api
    let appStore
    let authzStore
    let cloudProfileStore
    let gardenerExtensionStore
    let credentialStore

    beforeEach(async () => {
      setActivePinia(createPinia())
      api = useApi()
      vi.spyOn(api, 'getCloudProviderCredentials').mockImplementation(namespace => {
        if (namespace !== testNamespace) {
          throw new Error('Unauthorized')
        }
        return {
          data: fixtures.credentials,
        }
      })
      vi.spyOn(api, 'updateCloudProviderCredential').mockImplementation(({ secretBinding, secret }) => {
        return {
          data: {
            secretBinding,
            secret,
          },
        }
      })
      vi.spyOn(api, 'createCloudProviderCredential').mockImplementation(({ secretBinding, secret }) => {
        return {
          data: {
            secretBinding,
            secret,
          },
        }
      })
      vi.spyOn(api, 'deleteCloudProviderCredential').mockReturnValue()
      appStore = useAppStore()
      vi.spyOn(appStore, 'setSuccess')
      authzStore = useAuthzStore()
      authzStore.setNamespace(testNamespace)
      cloudProfileStore = useCloudProfileStore()
      cloudProfileStore.list = fixtures.cloudprofiles
      gardenerExtensionStore = useGardenerExtensionStore()
      gardenerExtensionStore.list = fixtures.gardenerExtensions
      credentialStore = useCredentialStore()
      credentialStore._setCredentials(fixtures.credentials)
    })

    afterEach(() => {
      appStore.setSuccess.mockClear()
      api.getCloudProviderCredentials.mockClear()
      api.createCloudProviderCredential.mockClear()
      api.updateCloudProviderCredential.mockClear()
      api.deleteCloudProviderCredential.mockClear()
    })

    it('should create a new credential store', () => {
      expect(cloudProfileStore.sortedProviderTypeList).toEqual(
        expect.arrayContaining(['aws', 'azure']),
      )
      expect(gardenerExtensionStore.dnsProviderTypes).toEqual(
        expect.arrayContaining(['aws-route53', 'azure-dns']),
      )
    })

    it('should return cloudProviderBindingList with resolved secret (secretbinding)', () => {
      const awsSecretBinding = find(credentialStore.cloudProviderBindingList, { metadata: { name: awsSecretBindingName } })
      expect(awsSecretBinding.secretRef.name).toBe(awsSecretBinding._secret.metadata.name)
      expect(awsSecretBinding.secretRef.namespace).toBe(awsSecretBinding._secret.metadata.namespace)
    })

    it('should return cloudProviderBindingList with resolved secret (credentialsbinding)', () => {
      const awsCredentialsBinding = find(credentialStore.cloudProviderBindingList, { metadata: { name: awsCredentialsBindingName } })
      expect(awsCredentialsBinding.credentialsRef.name).toBe(awsCredentialsBinding._secret.metadata.name)
      expect(awsCredentialsBinding.credentialsRef.namespace).toBe(awsCredentialsBinding._secret.metadata.namespace)
    })

    it('should return cloudProviderBindingList with trial quota and secret from other namespace (secretbinding)', () => {
      const awsTrialSecretBinding = find(credentialStore.cloudProviderBindingList, { metadata: { name: awsTrialSecretBindingName } })
      expect(awsTrialSecretBinding._secret).toBeUndefined()
      expect(awsTrialSecretBinding.secretRef.namespace).toBe(awsTrialSecretBinding._quotas[0].metadata.namespace)
    })

    it('should return cloudProviderBindingList with trial quota and secret from other namespace (credentialsbinding)', () => {
      const awsTrialCredentialsBinding = find(credentialStore.cloudProviderBindingList, { metadata: { name: awsTrialCredentialsBindingName } })
      expect(awsTrialCredentialsBinding._secret).toBeUndefined()
      expect(awsTrialCredentialsBinding.credentialsRef.namespace).toBe(awsTrialCredentialsBinding._quotas[0].metadata.namespace)
    })

    it('should return cloudProviderBindingList with multiple quotas (secretbinding)', () => {
      const azureSecretBinding = find(credentialStore.cloudProviderBindingList, { metadata: { name: azureSecretBindingName } })
      expect(azureSecretBinding._secret).toMatchSnapshot()
      expect(azureSecretBinding._quotas.length).toBe(2)
      const azureQuotaRef2 = azureSecretBinding.quotas[1]
      const azureQuota = find(credentialStore.quotaList, { metadata: azureQuotaRef2 })
      expect(azureQuota).toMatchSnapshot()
    })

    it('should return cloudProviderBindingList with multiple quotas (credentialsbinding)', () => {
      const azureCredentialsBinding = find(credentialStore.cloudProviderBindingList, { metadata: { name: azureCredentialsBindingName } })
      expect(azureCredentialsBinding._secret).toMatchSnapshot()
      expect(azureCredentialsBinding._quotas.length).toBe(2)
      const azureQuotaRef2 = azureCredentialsBinding.quotas[1]
      const azureQuota = find(credentialStore.quotaList, { metadata: azureQuotaRef2 })
      expect(azureQuota).toMatchSnapshot()
    })

    it('should return cloudProviderBindingList with resolved workloadidentity (credentialsbinding)', () => {
      const awsWorkloadIdentityCredentialsBinding = find(credentialStore.cloudProviderBindingList, { metadata: { name: awsWorkloadIdentitiyCredentialsBindingName } })
      expect(awsWorkloadIdentityCredentialsBinding.credentialsRef.name).toBe(awsWorkloadIdentityCredentialsBinding._workloadIdentity.metadata.name)
      expect(awsWorkloadIdentityCredentialsBinding.credentialsRef.namespace).toBe(awsWorkloadIdentityCredentialsBinding._workloadIdentity.metadata.namespace)
    })

    it('should return infrastructureBindingList', () => {
      expect(credentialStore.infrastructureBindingList.length).toBeGreaterThan(0)
      expect(credentialStore.cloudProviderBindingList.length).toBeGreaterThan(credentialStore.infrastructureBindingList.length)
      expect(cloudProfileStore.sortedProviderTypeList).toMatchSnapshot()
    })

    it('should return dnsBindingList', () => {
      expect(credentialStore.dnsBindingList.length).toBeGreaterThan(0)
      expect(credentialStore.cloudProviderBindingList.length).toBeGreaterThan(credentialStore.dnsBindingList.length)
      expect(gardenerExtensionStore.dnsProviderTypes).toMatchSnapshot()
    })

    it('should return secret', () => {
      expect(credentialStore.getSecret({ namespace: testNamespace, name: awsSecretName })).toMatchSnapshot()
    })

    it('should return workloadIdentity', () => {
      expect(credentialStore.getWorkloadIdentity({ namespace: testNamespace, name: awsWorkloadIndentityName })).toMatchSnapshot()
    })

    it('should return secretBinding', () => {
      expect(credentialStore.getSecretBinding({ namespace: testNamespace, name: awsSecretBindingName })).toMatchSnapshot()
    })

    it('should return credentialsBinding', () => {
      expect(credentialStore.getCredentialsBinding({ namespace: testNamespace, name: awsCredentialsBindingName })).toMatchSnapshot()
    })

    it('should fetchCredentials', async () => {
      credentialStore.$reset()
      expect(credentialStore.cloudProviderBindingList.length).toBe(0)
      await credentialStore.fetchCredentials()
      expect(api.getCloudProviderCredentials).toBeCalledTimes(1)
      expect(api.getCloudProviderCredentials).toBeCalledWith(testNamespace)
      expect(credentialStore.cloudProviderBindingList.length).toBeGreaterThan(0)
    })

    it('should updateCredential', async () => {
      let awsSecretBinding = find(credentialStore.cloudProviderBindingList, { metadata: { name: awsSecretBindingName } })
      const secret = {
        ...awsSecretBinding._secret,
        data: {
          newSecret2: 'dummy-data',
        },
      }
      await credentialStore.updateCredential({ secretBinding: awsSecretBinding, secret })

      expect(api.updateCloudProviderCredential).toBeCalledTimes(1)
      expect(api.updateCloudProviderCredential).toBeCalledWith({ secretBinding: awsSecretBinding, secret })
      awsSecretBinding = find(credentialStore.cloudProviderBindingList, { metadata: { name: awsSecretBindingName } })
      expect(awsSecretBinding._secret.data).toEqual({ newSecret2: 'dummy-data' })
    })

    it('should createCredential', async () => {
      const secretBinding = {
        apiVersion: 'core.gardener.cloud/v1alpha1',
        kind: 'SecretBinding',
        metadata: {
          namespace: testNamespace,
          name: 'my-new-secret-binding',
        },
        provider: {
          type: 'aws',
        },
        secretRef: {
          namespace: testNamespace,
          name: 'my-new-secret',
        },
      }
      const secret = {
        apiVersion: 'v1',
        kind: 'Secret',
        type: 'Opaque',
        metadata: {
          name: 'my-new-secret',
          namespace: testNamespace,
        },
        data: {
          newSecret: 'dummy-data',
        },
      }

      await credentialStore.createCredential({ secretBinding, secret })

      expect(api.createCloudProviderCredential).toBeCalledTimes(1)
      expect(api.createCloudProviderCredential).toBeCalledWith({ secretBinding, secret })
      const newSecretBinding = find(credentialStore.cloudProviderBindingList, { metadata: secretBinding.metadata })
      expect(newSecretBinding.metadata.namespace).toEqual(testNamespace)
      expect(newSecretBinding.provider.type).toEqual('aws')
      expect(newSecretBinding._secret.data).toEqual({ newSecret: 'dummy-data' })
    })

    it('should not delete credential', async () => {
      const name = azureSecretBindingName
      const namespace = testNamespace

      await credentialStore.deleteCredential(name)

      expect(api.deleteCloudProviderCredential).toBeCalledTimes(1)
      expect(api.deleteCloudProviderCredential).toBeCalledWith({ name, namespace })
      expect(api.getCloudProviderCredentials).toBeCalledTimes(1)
    })

    it('store should be resetted in case of a fetch error', async () => {
      const namespace = 'invalid'
      authzStore.setNamespace(namespace)

      expect(credentialStore.cloudProviderBindingList.length).toBeGreaterThan(0)
      await expect(credentialStore.fetchCredentials()).rejects.toThrow(Error)
      expect(credentialStore.cloudProviderBindingList.length).toBe(0)
    })
    it('store should be resetted after setting new data', async () => {
      const namespace = 'invalid'
      authzStore.setNamespace(namespace)

      expect(credentialStore.cloudProviderBindingList.length).toBeGreaterThan(0)
      credentialStore._setCredentials({})
      expect(credentialStore.cloudProviderBindingList.length).toBe(0)
    })
  })
})
