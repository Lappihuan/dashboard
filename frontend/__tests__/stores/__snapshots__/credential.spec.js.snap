// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`stores > credential > should return cloudProviderBindingList with multiple quotas (credentialsbinding) 1`] = `
{
  "data": {
    "secret": "cw==",
  },
  "metadata": {
    "name": "azure-secret",
    "namespace": "garden-test",
  },
}
`;

exports[`stores > credential > should return cloudProviderBindingList with multiple quotas (credentialsbinding) 2`] = `
{
  "metadata": {
    "name": "azure-bar-quota",
    "namespace": "garden-test",
  },
}
`;

exports[`stores > credential > should return cloudProviderBindingList with multiple quotas (secretbinding) 1`] = `
{
  "data": {
    "secret": "cw==",
  },
  "metadata": {
    "name": "azure-secret",
    "namespace": "garden-test",
  },
}
`;

exports[`stores > credential > should return cloudProviderBindingList with multiple quotas (secretbinding) 2`] = `
{
  "metadata": {
    "name": "azure-bar-quota",
    "namespace": "garden-test",
  },
}
`;

exports[`stores > credential > should return credentialsBinding 1`] = `
{
  "credentialsRef": {
    "kind": "Secret",
    "name": "aws-secret",
    "namespace": "garden-test",
  },
  "metadata": {
    "name": "aws-credentialsbinding",
    "namespace": "garden-test",
  },
  "provider": {
    "type": "aws",
  },
}
`;

exports[`stores > credential > should return dnsBindingList 1`] = `
[
  "aws-route53",
  "azure-dns",
  "azure-private-dns",
  "google-clouddns",
  "openstack-designate",
  "alicloud-dns",
  "infoblox-dns",
  "netlify-dns",
  "rfc2136",
]
`;

exports[`stores > credential > should return infrastructureBindingList 1`] = `
[
  "aws",
  "azure",
  "gcp",
  "openstack",
  "alicloud",
  "ironcore",
]
`;

exports[`stores > credential > should return secret 1`] = `
{
  "data": {
    "secret": "cw==",
  },
  "metadata": {
    "name": "aws-secret",
    "namespace": "garden-test",
  },
}
`;

exports[`stores > credential > should return secretBinding 1`] = `
{
  "metadata": {
    "name": "aws-secretbinding",
    "namespace": "garden-test",
  },
  "provider": {
    "type": "aws",
  },
  "secretRef": {
    "name": "aws-secret",
    "namespace": "garden-test",
  },
}
`;

exports[`stores > credential > should return workloadIdentity 1`] = `undefined`;
